android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java

```diff
@@ -652,6 +671,74 @@ private String encodeChannelEvent(ChannelEvent event) {
     return jsonMap;
   }
 
+  private Map<String, Object> encodeDeviceDetails(DeviceDetails c) {
+    if (c == null) return null;
+    final HashMap<String, Object> jsonMap = new HashMap<>();
+
+    writeValueToJson(jsonMap, PlatformConstants.TxDeviceDetails.id, c.id);
+    writeValueToJson(jsonMap, PlatformConstants.TxDeviceDetails.clientId, c.clientId);
+    writeValueToJson(jsonMap, PlatformConstants.TxDeviceDetails.platform, c.platform);
+    writeValueToJson(jsonMap, PlatformConstants.TxDeviceDetails.formFactor, c.formFactor);
+    writeValueToJson(jsonMap, PlatformConstants.TxDeviceDetails.metadata, c.metadata);
+    writeValueToJson(jsonMap, PlatformConstants.TxDeviceDetails.devicePushDetails, encodeDevicePushDetails(c.push));
+
+    return jsonMap;
+  }
+
+  private Map<String, Object> encodeDevicePushDetails(DeviceDetails.Push c) {
+    if (c == null) return null;
+    final HashMap<String, Object> jsonMap = new HashMap<>();
+
+    writeValueToJson(jsonMap, PlatformConstants.TxDevicePushDetails.recipient, c.recipient);
+    writeValueToJson(jsonMap, PlatformConstants.TxDevicePushDetails.state, encodeDevicePushDetailsState(c.state));
+    writeValueToJson(jsonMap, PlatformConstants.TxDevicePushDetails.errorReason, encodeErrorInfo(c.errorReason));
+
+    return jsonMap;
+  }
+
+  private String encodeDevicePushDetailsState(DeviceDetails.Push.State state) {
+    if (state == null) return null;
+
+    switch (state) {
+      case ACTIVE:
+        return PlatformConstants.TxDevicePushStateEnum.active;
+      case FAILING:
+        return PlatformConstants.TxDevicePushStateEnum.failing;
+      case FAILED:
+        return PlatformConstants.TxDevicePushStateEnum.failed;
+      default:
+        return null;
```

Comment:
Is this ever a valid situation? I would rather you failed and threw an exception if this is indicative of an issue elsewhere.

Reply:
That's a good point, and I would prefer it, but I was just following the rest of the file's style, `return null` occurs 42 times in that file (trying to be consistent).

Reply:
I think we can track this in a separate issue to throw a platform exception instead of returning null.

Reply:
Okay, I created https://github.com/ably/ably-flutter/issues/149